### PR TITLE
v4

### LINKED ISSUE #490: Support CSS variables or SASS export
Currently, the neetoUI SASS theme variables and mixins have scope only inside the neetoUI.  
If we need a mixin or theme variable on the project side we need to copy-paste it from neetoUI.

If we support SASS export or CSS variables, we can use the CSS variables and mixins we define in neetoUI on the project side also.

Another advantage is we can create different themes based on the project with minimal effort. 

Let’s consider this once we release a stable version of neetoV2.

### LINKED ISSUE #504: Consistency Improvements
@praveen-murali-ind _a

### Single Source of Truth - Design System

A few suggestions/thoughts to improve the UI consistency in applications using neetoUI
(V2).

Currently, we are using customised Tailwind in projects.

If we have 100 applications, there will be 100 different Tailwind configurations, so the Tailwind utility classes may be different while moving from one project to another even if we are using the same Tailwind version.

**1. We should stop customising Tailwind themes and use the defaults provided by Tailwind**

_Advantages_

- Tailwind utility classes would be the same across all projects and developers will have a better idea of which all classes are available.
-  Easy to upgrade Tailwind version.

**2. For theme-related CSS properties like `color`, `background-color`, `border-radius`, `box-shadow`, `border-color`, etc, we should stop using Tailwind palette and utility classes and we should use neetoUI Design System palette and utility classes.**

[neetoUI helper classes](https://neetoui-v2.netlify.app/?path=/story/foundation-helpers-border-color--page)

Rather than creating neetoUI like palette using Tailwind on the project side, or using similar colors, we should use the exact neetoUI palette.

_Advantages_

- The design system we created could act as a single source of truth for all projects.
- We can ensure consistency throughout applications.
- We will have more control over the UI.
- For other CSS properties, we can take advantage of Tailwind utility classes.

**3. Until we support neetoUI SASS import, let’s create a file named `neeto-ui-variables.scss` in the project to clone the neetoUI SASS `variables` and `mixins`. Don’t include any project-specific variables in this file. We will be removing this once we support SASS import in neetoUI.**

Example Tailwind config file in project👇

Screenshot 2021-10-08 at 6 40 36 PM

### LINKED ISSUE #772: Fix delay while opening Modal
@praveen-murali-ind
The customer creation modal is having a noticeable amount of frame drop. Is it fixable?

https://user-images.githubusercontent.com/85148587/150116679-0fbd3524-2d44-4323-b35b-3cdb1a0d98c5.mov

_Originally posted by @Amaljith-K in https://github.com/bigbinary/neeto-chat-web/issues/2089#issuecomment-1016328480_

### LINKED ISSUE #1182: Real-time theme customization using CSS variables
@praveen-murali-ind _a

- [x] Include CSS variables to enable real-time customization and easier theming

CSS variables - Reference guide

```
:root {
    --neeto-ui-black: #1b1f23;
    --neeto-ui-white: #ffffff;
    --neeto-ui-gray-800: #2f3941;
    --neeto-ui-gray-700: #49545c;
    --neeto-ui-gray-600: #68737d;
    --neeto-ui-gray-500: #87929d;
    --neeto-ui-gray-400: #c2c8cc;
    --neeto-ui-gray-300: #d8dcde;
    --neeto-ui-gray-200: #e9ebed;
    --neeto-ui-gray-100: #f8f9f9;
    --neeto-ui-success: #00ba88;
    --neeto-ui-info: #276ef1;
    --neeto-ui-warning: #f57c00;
    --neeto-ui-error: #ea4335;
    --neeto-ui-pastel-blue: #eaf3fc;
    --neeto-ui-pastel-green: #ebf5ec;
    --neeto-ui-pastel-yellow: #fff2d7;
    --neeto-ui-pastel-red: #ffefed;
    --neeto-ui-pastel-teal: #98f3f4;
    --neeto-ui-accent1-700: #5b5ac4;
    --neeto-ui-accent2-700: #00956d;
    --neeto-ui-accent3-700: #50a8cc;
    --neeto-ui-accent4-700: #d62516;
    --neeto-ui-accent1-500: #5e5ce6;
    --neeto-ui-accent2-500: #00ba88;
    --neeto-ui-accent3-500: #64d2ff;
    --neeto-ui-accent4-500: #ea4335;
    --neeto-ui-accent1-300: #7e7deb;
    --neeto-ui-accent2-300: #33c8a0;
    --neeto-ui-accent3-300: #83dbff;
    --neeto-ui-accent4-300: #f78879;
    --neeto-ui-secondary-indigo: #5e5ce6;
    --neeto-ui-secondary-green: #00ba88;
    --neeto-ui-secondary-teal: #64d2ff;
    --neeto-ui-shadow-xs: 0px 1px 4px -1px rgba(28, 48, 74, 0.12);
    --neeto-ui-shadow-s: 0px 3px 12px -1px rgba(28, 52, 84, 0.12), 0px 2px 4px -1px rgba(28, 55, 90, 0.08);
    --neeto-ui-shadow-m: 0px 8px 24px -4px rgba(28, 50, 79, 0.12), 0px 2px 6px -1px rgba(28, 55, 90, 0.08);
    --neeto-ui-shadow-l: 0px 12px 48px -6px rgba(28, 50, 79, 0.12), 0px 3px 18px -2px rgba(28, 55, 90, 0.08);
    --neeto-ui-shadow-xl: 0px 18px 64px -8px rgba(28, 50, 79, 0.12), 0px 4px 24px -3px rgba(28, 55, 90, 0.08);
    --neeto-ui-text-h1: 32px;
    --neeto-ui-text-h2: 24px;
    --neeto-ui-text-h3: 20px;
    --neeto-ui-text-h4: 16px;
    --neeto-ui-text-h5: 14px;
    --neeto-ui-text-h6: 12px;
    --neeto-ui-text-body1: 16px;
    --neeto-ui-text-body2: 14px;
    --neeto-ui-text-body3: 12px;
    --neeto-ui-text-xxs: 10px;
    --neeto-ui-text-xs: 12px;
    --neeto-ui-text-sm: 14px;
    --neeto-ui-text-base: 16px;
    --neeto-ui-text-lg: 18px;
    --neeto-ui-text-xl: 20px;
    --neeto-ui-text-2xl: 24px;
    --neeto-ui-text-3xl: 32px;
    --neeto-ui-text-4xl: 48px;
    --neeto-ui-font-thin: 100;
    --neeto-ui-font-extralight: 200;
    --neeto-ui-font-light: 300;
    --neeto-ui-font-normal: 400;
    --neeto-ui-font-medium: 500;
    --neeto-ui-font-semibold: 600;
    --neeto-ui-font-bold: 700;
    --neeto-ui-font-extrabold: 800;
    --neeto-ui-font-black: 900;
    --neeto-ui-leading-none: 1;
    --neeto-ui-leading-tight: 1.25;
    --neeto-ui-leading-snug: 1.375;
    --neeto-ui-leading-normal: 1.5;
    --neeto-ui-leading-relaxed: 1.625;
    --neeto-ui-leading-loose: 2;
    --neeto-ui-text-transform-none: none;
    --neeto-ui-text-transform-capitalize: capitalize;
    --neeto-ui-text-transform-uppercase: uppercase;
    --neeto-ui-text-transform-lowercase: lowercase;
    --neeto-ui-text-transform-full-width: full-width;
    --neeto-ui-text-transform-inherit: inherit;
    --neeto-ui-text-transform-initial: initial;
    --neeto-ui-text-transform-revert: revert;
    --neeto-ui-text-transform-unset: unset;
    --neeto-ui-rounded-none: 0;
    --neeto-ui-rounded-sm: 2px;
    --neeto-ui-rounded: 4px;
    --neeto-ui-rounded-md: 6px;
    --neeto-ui-rounded-lg: 8px;
    --neeto-ui-rounded-xl: 10px;
    --neeto-ui-rounded-full: 999px;
    --neeto-ui-transition: all 0.3s ease-in-out;
    --neeto-ui-pane-header-height: 78px;
    --neeto-ui-pane-footer-height: 80px;
    --neeto-ui-main-header-height: 80px;
    --neeto-ui-sub-header-height: 37px;
    --neeto-ui-sub-header-bottom-margin: 20px;
    --neeto-ui-pagination-block-height: 88px;
    --neeto-ui-table-header-height: 40px;
    --neeto-ui-modal-z-index: 99999;
}
```

### LINKED ISSUE #1183: Replace Framer Motion animations with CSS animations
@praveen-murali-ind _a

- [x] Menubar
- [x] Sidebar
- [x] AppSwitcher
- [x] Modal
- [x] Alert
- [x] Pane

### LINKED ISSUE #1200: Make text in app switcher more visible
Ref: https://github.com/bigbinary/neeto-auth-web/issues/1671

### LINKED ISSUE #1205: Cosmetic changes
@praveen-murali-ind _a

Form Elements

- [x] Button - #1204 
- [x] Checkbox - #1207
- [x] Radio - #1208
- [x] Input - #1209
- [x] Textarea - #1210
- [x] Select - #1215
- [x] Email Input - #1211
- [x] Label - #1212
- [x] Dropdown - #1216
- [x] Action Dropdown - #1217
- [x] Date and Time - #1213
- [x] Accordion - #1218
- [x] Avatar - #1219
- [x] Table - #1220
- [x] Pagination - #1222
- [x] PageLoader - #1230
- [x] Switch - no changes
- [x] Tag - no changes
- [x] Tab - no changes
- [x] Color Picker -  #1235 
- [x] Spinner - no changes
- [x] Callout - no changes

Overlays

- [x] Alert - #1227
- [x] Modal - #1223
- [x] Toastr - #1228
- [x] Tooltip - #1229
- [x] Pane - no changes

Layouts

- [x] AppSwitcher - pending
- [x] Menubar - 
- [x] Sidebar - #1231
- [x] Header - #1234

### PR DESCRIPTION
**Pull request by praveen-murali-ind**
Fixes #1205 
Fixes #490 
Fixes #504
Fixes #772  
Fixes #1183 
Fixes #1182 
Fixes #1200 

**Description**

- Updated: Color palette
- Added: CSS variables support
- Added: Custom theming support
- Added: Default dark theme support
- Added: Design preview in Storybook
- Added: Guidelines to structure content in applications
- Changed: bundle type to `CJS`
- Fixed: delay while opening overlays
- Removed: `framer-motion` from dependencies.

### Color palette 🎨 

- Updated color naming convention
- Added CSS variables support for colors
- Implemented dynamic color palette to support default dark mode
- Served color codes as RBG numbers in CSS variables to achieve color transparency
- Enhanced contrast of status color (info, success, error, warning)
- Included existing pastel colors as the color swatches of primary, success, error, warning, and info
- Added new set of pastel colors
- Removed unused colors from the palette
- Updated utility classes to match the new variable naming
- Restructured documentation and created new documentation for color palette
- Deprecated `neeto-ui-accent1`, `neeto-ui-accent2`, `neeto-ui-accent3`, `neeto-ui-accent4` colors and utility classes
- Deprecated `neeto-ui-secondary-indigo`, `neeto-ui-secondary-green`, `neeto-ui-secondary-teal` colors and utility classes
- Deprecated `neeto-ui-pastel-teal`
- Renamed `neeto-ui-success` to `neeto-ui-success-500`
- Renamed `neeto-ui-info` to `neeto-ui-info-500`
- Renamed `neeto-ui-error` to `neeto-ui-error-500`
- Renamed `neeto-ui-warning` to `neeto-ui-warning-500`
- Renamed `neeto-ui-pastel-blue` to `neeto-ui-info-100`
- Renamed `neeto-ui-pastel-green` to `neeto-ui-success-100`
- Renamed `neeto-ui-pastel-yellow` to `neeto-ui-warning-100`
- Renamed `neeto-ui-pastel-red` to `neeto-ui-error-100`

#### Button

- Updated primary button color to `neeto-ui-primary-500`
- Added `:focus` rings to buttons for accessibility
- Updated `:focus-visible` styles to match the new design 
- Updated border radius to `5px`
- Updated CSS transition from `all .3s ease-in-out` to `all .3s linear`
- Deprecated variants story in Storybook doc
- Created independent stories for `Styles`, `Sizes`, `Icon only` etc

#### Checkbox

- Replaced checked state icon with neetoIcon
- Updated checked state color to `neeto-ui-primary-500`
- Changed default state border color from `grey 300` to `grey 400`
- Updated `:hover` border color to `grey 500`
- Updated `:focus-visible` border color to `neeto-ui-primary-500`
- Updated focus ring styles to match `neeto-ui-primary` color
- Improved spacing in component

#### Radio

- Replaced checked state icon with neetoIcon
- Updated checked state color to `neeto-ui-primary-500`
- Changed default state border color from `grey 300` to `grey 400`
- Updated `:hover` border color to `grey 500`
- Updated `:focus-visible` border color to `neeto-ui-primary-500`
- Updated focus ring styles to match `neeto-ui-primary` color
- Improved spacing in component

#### Input 

- Updated border-radius to `5px`
- Updated focus ring styles to match `neeto-ui-primary` color
- Fixed unwanted spacing in maxLength label 
- Fixed alignment issues in Input with maxLength and no label text
- Fixed spacing inconsistencies in error test and helper text - used same margin value for both
- Fixed alignment issues in naked inputs
- Deprecate `variants` story in Storybook and created individual stories for `Input sizing`

#### Textarea

- Updated border-radius to `5px`
- Updated focus ring styles to match `neeto-ui-primary` color
- Fixed unwanted spacing in the maxLength label 
- Fixed alignment issues in Input with maxLength and no label text
- Fix spacing inconsistencies in error test and helper text - used the same margin value for both
- Fixed alignment issues in naked inputs
- Added size prop to Textarea

#### Select

- Updated border-radius to `5px`
- Updated focus ring styles to match `neeto-ui-primary` color
- Replaced the default dropdown icons with neetoIcons
- Added CSS transition to the select options
- Applied Badge styles to Multi select items
- Replaced multivalue remove icon with neetoIcon
- Updated multivalue error styles
- Increased Story heights for better view

#### Email input

- Updated border radius to `5px`
- Updated selected values styles to match Tag styles
- Fixed prefix and suffix spacing
- Updated prefix icon size to `16px`
- Updated focus ring styles to match `neeto-ui-primary` color
- Fixed spacing issues in counter 
- Fixed alignment issues in label row
- Fixed spacing inconsistencies in error test and helper text - used same margin for both
- Updated error styles of multivalue select

#### Label

- Updated label bottom margin to 8px
- Removed letter-spacing in label text

#### Dropdown & Action Dropdown

- Updated border-radius to `5px`
- Reduced the height of Dropdown options to `32px`
- Created Storybook examples with semantic markup for easy keyboard win
- Separated Delete/Remove actions with a border
- Updated `box-shadow`

#### Date Time picker

- Replaced default icons with neetoIcons
- Updated selection color to `neeto-ui-primary`
- Added border to date and time dropdown
- Updated box shadow of dropdown

#### Accordion

- Updated font weight of Accordion title to `500`
- Updated `border-radius` to `8px`

#### Avatar

- Updated background colors to new pastel colors
- Updated square variant `border-radius` to `5px`

#### Table

- Updated column title color to `grey 600`
- Updated column title font weight to 700
- Reduced cell padding from `12px 16px` to `8px 16px` to accommodate more number of row in a given height 
- Removed border from pagination items
- Updated pagination buttons to match neetoUI button style
- Updated active pagination item styles to `neeto-ui-primary`
- Replaced default icons with neetoIcons

#### PageLoader

- Updated neeto logo

#### Color picker

- Added border to input

#### Alert & Modal

- Reduced outer padding from `32px ` to `24px` to provide the compact view
- Updated `border-radius` to `12px`

#### Toast

- Applied `48px` minimum height
- Applied dark theme to all variants
- Reduced text font weight to `400`

#### Tooltip

- Changed box shadow of light theme variant to `--neeto-ui-shadow-s`

#### Sidebar

- Increased width from `64px` to `72px`
- Changed background color from `neeto-ui-gray-200` to `neeto-ui-white`
- Added border on the right side
- Updated `padding-top` and `padding-bottom` to `24px`

#### Input

- Changed: `maxlength` prop logic to allow typing characters beyond the specified maxlength. The character count will be shown only when the input length reaches 90% of maxlength and changes the state to error when input exceeds the maxlength

#### TextArea

- Changed: `maxlength` prop logic to allow typing characters beyond the specified maxlength. Character count will be shown only when the input length reaches 90% of maxlength and changes the state to error when input exceeds the maxlength

#### Dropdown

- Changed: **BREAKING** Renamed `ulProps` prop to `dropdownProps`
- Added: `buttonSize` prop

#### ActionDropdown

- Changed: **BREAKING** Renamed `size` prop to `buttonSize`
- Changed: **BREAKING** Renamed `style` prop to `buttonStyle`

#### Tag

- Changed: **BREAKING** Renamed `style` prop to `type`
- Removed: **BREAKING** Renamed `color` and `indicatorColor` props
- Added: **BREAKING** `style` and `indicatorStyle` props with options `"success" | "warning" | "danger" | "primary" | "inactive"`

#### ColorPicker

- Added: `Size` prop

#### Typography

- Added: new variant **nano** 

#### Modal

- Added: `description` prop in header
- Added: `initialFocusRef` and `finalFocusRef` to programmatically add focus to custom components.
- Added: `blockScrollOnMount` prop to enable/disable body scrolling when the modal opens.
- Added: focus trapping to focusable elements in Modal.

#### Sidebar

- Removed: **BREAKING** `footerLinks` props
- Removed: **BREAKING** `isCollapsed` prop and **expanded state** of sidebar

[Storybook Documentation](https://6224dd6a6655cb003a07b79c-tlniiktvyu.chromatic.com/) 

**Checklist**

- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary label (patch/minor/major - If package publish is required)
- [ ] I have followed the suggested description format and styling

**Reviewers**

### CODE REVIEW

**Comment by praveen-murali-ind** [comment]
@praveen-murali-ind _a

**Comment by praveen-murali-ind** [comment]
@amaldinesh7 _a

File: lib/components/Button.js [hunk 1]
```diff
@@ -15,132 +15,135 @@ const BUTTON_STYLES = {
...
+              /* When Icon is not present, animate the margin from 0 to the needed value*/
+              loading && (
+                <motion.div
+                  className="neeto-ui-btn__spinner-wrapper"
+                  initial={{ width: 0, scale: 0 }}
+                  animate={{
+                    width: "auto",
+                    scale: 1,
+                  }}
+                  exit={{ width: 0, scale: 0 }}
```
Comment (ajmaln) [inline lib/components/Button.js:129]: Can we move these to variants?

File: tests/Tag.test.js [hunk 2]
```diff
@@ -14,8 +15,8 @@ describe("Tag", () => {
     expect(getByTestId("class-icon")).toBeInTheDocument();
   });
 
-  it("should show indicator when indicatorColor is provided", () => {
-    const { getByTestId } = render(<Tag indicatorColor="green" />);
+  it("should show indicator when indicatorStatus is provided", () => {
+    const { getByTestId } = render(<Tag indicatorStyle="success" />);
```
Comment (ajmaln) [inline tests/Tag.test.js:19]:
```suggestion
  it("should show indicator when indicatorStyle is provided", () => {
    const { getByTestId } = render(<Tag indicatorStyle="success" />);
```

**Review by ajmaln — APPROVED** [review]
@amaldinesh7 _a Few minor suggestions, rest LGTM

**Review by amaldinesh7 — APPROVED** [review]

**Review by karthiknmenon — APPROVED** [review]

**Pull request merged**